cinn/optim/vectorize_loops.cc

```diff
@@ -50,6 +55,177 @@ Expr Widen(Expr e, int lanes) {
+// of tensors which doesn't meet all predicates of vectoring
+class TensorVectorizeExcluder : public ir::IRMutator<const Expr *> {
+ public:
+  TensorVectorizeExcluder(const std::string &iter_name) : iter_name_(iter_name) {}
+
+  void Collect(const Expr *op) { IRMutator::Visit(op, op); }
+
+  bool Contain(const std::string &tensor_name) const { return excluded_tensors_.count(tensor_name); }
+
+ private:
+  const std::string iter_name_;
+  std::set<std::string> excluded_tensors_;
+
+  void Visit(const ir::Store *expr, const Expr *op) override {
+    auto *node = op->As<ir::Store>();
+    CHECK(node);
+    IRMutator::Visit(&node->value, &node->value);
+    auto *tensor = node->tensor.As<ir::_Tensor_>();
+    CHECK(tensor);
+    if (HitBan(node->tensor, node->indices)) excluded_tensors_.insert(tensor->name);
+  }
+
+  void Visit(const ir::Load *expr, const Expr *op) override {
+    auto *node = op->As<ir::Load>();
+    CHECK(node);
+    auto *tensor = node->tensor.As<ir::_Tensor_>();
+    CHECK(tensor);
+    if (HitBan(node->tensor, node->indices)) excluded_tensors_.insert(tensor->name);
+  }
+
+  // return true if the tensor hit some prohibitions of vectorizing
+  bool HitBan(Expr tensor, const std::vector<Expr> &indices) {
+    auto find_matched_var_fn = [&](const Expr *x) { return x->As<_Var_>() && x->As<_Var_>()->name == iter_name_; };
+    auto in_last_index       = ir::CollectIRNodes(indices.back(), find_matched_var_fn);
+
+    // the iter val must appear in the last index
+    if (in_last_index.empty()) {
+      VLOG(5) << "Loop var:" << iter_name_
+              << " is not used in the last index, so the memory access will be not sequential";
+      return true;
+    }
+
+    // the iter val can't appear in mulitple indices
+    for (int i = 0; i < indices.size() - 1; ++i) {
+      auto repeat_found = ir::CollectIRNodes(indices[i], find_matched_var_fn);
+      if (!repeat_found.empty()) {
+        VLOG(5) << "Loop var:" << iter_name_ << " is also used at the index:" << i;
+        return true;
+      }
+    }
+
+    auto dtype = tensor->type().ElementOf();
+    if (dtype.is_float(32) || dtype.is_int(32) || dtype.is_uint(32)) {
+      return false;
+    }
+    VLOG(5) << "Only support vectorizing int,uint,float";
+    return true;
+  }
+};
+
+// find tensors accessed sequentially in a for-loop to be vectorized,
+// and substitue the corresponding cuda built-in vector for them
+class CudaVectorizer : public IRMutator<Expr *> {
+  const Var iter_var_;
+  const int factor_;
+
+  TensorWriteTeller write_teller_;
+  TensorVectorizeExcluder tensor_excluder_;
+
+  absl::flat_hash_map<std::string, Var> tensor2vectorized_vars_;
+  std::vector<Expr> vectorized_cast_exprs_;
+
+ public:
+  CudaVectorizer(const Var &iter_var, const int factor)
+      : iter_var_(iter_var), factor_(factor), tensor_excluder_(iter_var->name) {
+    CHECK(factor == 2 || factor == 4) << "Only support factor 2 or 4 in CUDA type vectorizing, input factor:" << factor;
```

Comment:
Can we make these `2`, `4` some kind of constant to be maintainable in the future?

Reply:
fix it, define the maximum lane of CUDA vector types as a  static const number and update this check